include/lsst/afw/table/python/catalog.h

```diff
@@ -21,102 +19,117 @@
-    });
-    cls.def(("_lower_bound_" + suffix).c_str(),
-            [](Catalog & self, T const & value, Key<T> const & key)->int {
-        return self.lower_bound(value, key) - self.begin();
-    });
-    cls.def(("_equal_range_" + suffix).c_str(),
-            [](Catalog & self, T const & value, Key<T> const & key)->std::pair<int,int> {
-        std::pair<typename Catalog::const_iterator, typename Catalog::const_iterator> p
-            = self.equal_range(value, key);
-        return std::pair<int,int>(p.first - self.begin(), p.second - self.begin());
-    });
-};
+    cls.def("sort", (void (Catalog::*)(Key<T> const &)) &Catalog::sort);
+    cls.def(
+        "find",
+        [](Catalog & self, Value const & value, Key<T> const & key) -> std::shared_ptr<Record>
+        {
+            auto iter = self.find(value, key);
+            if (iter == self.end()) {
+                return nullptr;
+            };
+            return iter;
+        }
+    );
+    cls.def(
+        "upper_bound",
+        [](Catalog & self, Value const & value, Key<T> const & key) -> std::ptrdiff_t {
+            return self.upper_bound(value, key) - self.begin();
+        }
+    );
+    cls.def(
+        "lower_bound",
+        [](Catalog & self, Value const & value, Key<T> const & key) -> std::ptrdiff_t {
+            return self.lower_bound(value, key) - self.begin();
+        }
+    );
+    cls.def(
+        "equal_range",
+        [](Catalog & self, Value const & value, Key<T> const & key) {
+            auto p = self.equal_range(value, key);
+            return py::slice(p.first - self.begin(), p.second - self.begin(), 1);
+        }
+    );
+    cls.def(
+        "between",
+        [](Catalog & self, Value const & lower, Value const & upper, Key<T> const & key) {
+            std::ptrdiff_t a = self.lower_bound(lower, key) - self.begin();
+            std::ptrdiff_t b = self.upper_bound(upper, key) - self.begin();
+            return py::slice(a, b, 1);
+        }
+    );
+}
 
 /**
-Declare member and static functions for a given instantiation of lsst::afw::table::CatalogT<RecordT>.
+Wrap an instantiation of lsst::afw::table::CatalogT<Record>.
 
-@tparam RecordT  Record type, e.g. BaseRecord or SimpleRecord.
+In addition to calling this method you must call addCatalogMethods on the
+class object in Python.
 
-@param[in] cls  Catalog pybind11 class.
+@tparam Record  Record type, e.g. BaseRecord or SimpleRecord.
+
+@param[in] mod    Module object class will be added to.
+@param[in] name   Name prefix of the record type, e.g. "Base" or "Simple".
+@param[in] isBase Whether this instantiation is only being used as a base class (used to set the class name).
 */
-template <typename RecordT>
-void declareCatalog(pybind11::class_<CatalogT<RecordT>, std::shared_ptr<CatalogT<RecordT>>> & cls) {
+template <typename Record>
+PyCatalog<Record> declareCatalog(pybind11::module & mod, std::string const & name, bool isBase=false) {
```

Comment:
Perhaps (in the sense that I've used this on all the `Catalog` declare functions), but the C++ pattern it's wrapping (multiple levels of CRTP base classes) is not a common one, and in fact one I somewhat regret and would not like to see used again.